src/librustc_resolve/late.rs

```diff
@@ -544,6 +544,50 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
             self.visit_where_predicate(p);
         }
     }
+
+    fn visit_generic_arg(&mut self, arg: &'tcx GenericArg) {
+        debug!("visit_generic_arg({:?})", arg);
+        match arg {
+            GenericArg::Type(ref ty) => {
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    // A path type argument could be an incorrectly parsed generic const argument.
+                    // We double check just to be sure by trying to resolve in the type namespace
+                    // first, and if that fails we try again in the value namespace. If resolution
+                    // in the value namespace succeeds, we have an improperly parsed generic const
+                    // argument on our hands.
+                    if path.segments.len() == 1 {
+                        if let None = self.resolve_qpath(
+                            ty.id,
+                            qself.as_ref(),
+                            &Segment::from_path(path),
+                            Namespace::TypeNS,
+                            ty.span,
+                            CrateLint::SimplePath(ty.id),
+                        ) {
+                            if let Some(partial_res) = self.resolve_qpath(
+                                ty.id,
+                                qself.as_ref(),
+                                &Segment::from_path(path),
+                                Namespace::ValueNS,
+                                ty.span,
+                                CrateLint::SimplePath(ty.id),
+                            ) {
+                                debug!(
+                                    "visit_generic_arg: generic type argument resolved as a const"
+                                );
+                                self.r.record_partial_res(ty.id, partial_res);
+                                return;
```

Comment:
Forgetting to visit generic arguments here.
That may be the reason for the strange behavior in tests where `{}`s could not be removed due to type errors.

Reply:
The non-type case should generally mirror treatment of `ExprKind::Path` in `resolve_expr`.